// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";

package optimization_guide.proto.features;

import "components/optimization_guide/proto/features/forms_predictions.proto";

import "components/optimization_guide/proto/features/common_quality_data.proto";

import "components/optimization_guide/proto/common_types.proto";

import "components/optimization_guide/proto/model_quality_metadata.proto";

option optimize_for = LITE_RUNTIME;
option java_package = "org.chromium.components.optimization_guide.features.proto";

option java_outer_classname = "ModelPrototypingProto";

// DO NOT EDIT THIS FILE DIRECTLY!
//
// This file is generated in g3 and then synced to Chrome. Instead, please
// refer to http://go/chrome-intelligence-feature-protos (Google-internal link),
// and then changes will be synced with Chrome automatically.

message ModelPrototypingLoggingData {
  ModelPrototypingRequest request = 1;

  ModelPrototypingResponse response = 2;

  ModelPrototypingMetadata metadata = 3;

  ModelExecutionInfo model_execution_info = 6;
}

message ModelPrototypingRequest {
  ModelingInputs modeling_inputs = 1;

  // The series of prompts to send to the model(s). The calls are run in series
  // and the responses can be used in future calls allowing piping the output of
  // one query into the input of the next.
  repeated PrototypingPrompt prototyping_prompts = 2;

  // The responses from previous calls to the model. Can be used in future
  // prompts. Syntax for accessing them is golang text/templates
  // e.g., something like {{index .GetModelResponses 0}}.
  repeated string model_responses = 3;

  // Defines a single prompt to be sent to the model.
  message PrototypingPrompt {
    // Prompt variables that can be used in the rest of the prompt. These are in
    // addition to any prompt variables defined in the prompt template in the
    // config for the model sequence. Prompt variables are helper functions that
    // can be used in the prompt. For example, a prompt variable could be
    // something like:
    // {{ $funVar := "1" }}
    // This would define a function that can be used in the prompt as
    // {{$funVar}}. The value of the function is "1".
    string prompt_variables = 1;

    // The prompt is composed by inserting the following roles into the prompt
    // template in the order they are defined.
    // Role system is generally the instructions for the model to follow.
    string system_instructions_template = 2;

    // Role context is the information around the user interaction such as page
    // state.
    string context_area_template = 3;

    // Role user is the information from the user such as a user input they
    // typed.
    string user_input_template = 4;

    // Information about the model to use.
    ModelInformation model_information = 5;

    // The image to use in the prompt.
    string image_template = 6;

    // The pre-existing conversation between the user and the model.
    Conversation conversation = 7;

    message ModelInformation {
      ModelEnum model_enum = 1;

      bool skip_model = 2;

      enum ModelEnum {
        MODEL_ENUM_UNSPECIFIED = 0;

        MODEL_ENUM_EVERGREEN_GEMINI_V2 = 1;

        MODEL_ENUM_SUP_GEMINI_V2 = 2;

        MODEL_ENUM_EVERGREEN_GEMINI_V2_MULTI_MODAL = 3;

        MODEL_ENUM_EVERGREEN_GEMINI_V2_MULTI_MODAL_MPP = 4;
      }
    }
  }

  // Data specific to the feature.
  message ModelingInputs {
    BrowserCollectedInformation browser_collected_information = 1;

    string user_input = 2;
  }
}

// A representation of an ongoing conversation.
message Conversation {
  // The turns of various messages in the conversation.
  repeated ConversationTurn conversation_turns = 1;
}

// A single turn in a conversation.
message ConversationTurn {
  // The text of the conversation turn.
  string conversation_text = 1;

  // Which participant in the conversation this turn is from.
  string role = 2;

  // The mime type of the conversation turn.
  string mime_type = 3;
}

// All the information collected from the browser along with the user input
// (for features like Compose).
// NEXT ID: 14
message BrowserCollectedInformation {
  // The page context of the page the model is acting on.
  PageContext page_context = 1;

  // Deprecated, but still sent by older clients. Use page_context instead.
  string inner_text = 2;

  // Deprecated, but still sent by older clients. Use page_context instead.
  uint64 inner_text_offset = 3;

  // Custom text that a prototyper can inject into prompts. If the browser
  // collected information is not sufficient, an early stage prototype can
  // build a string in Chrome/colab to be used in the prompt. This allows
  // separation of prompt definition and call specific data.
  repeated string custom_data = 4;

  // The form to predict values for. It is expected for the values for each
  // field to be empty.
  FormData form_data = 5;

  // The current entries in the user annotations database helpful for filling
  // out this form.
  repeated UserAnnotationsEntry entries = 6;

  // The ungrouped tabs that are requested to be organized.
  repeated Tab tabs = 7;

  // The tab that was active at the time the user requested tab organization.
  int64 active_tab_id = 8;

  // All pre-existing tab groups.
  repeated TabGroup pre_existing_tab_groups = 9;

  // The site engagement information for the user.
  SiteEngagement site_engagement = 10;

  // The recent historical page data.
  repeated HistoryQueryResult history_query_result = 11;

  // Forms predictions metadata used for data collection/evaluations.
  FormsPredictionsRequest forms_predictions_request = 12;

  FormsPredictionsResponse forms_predictions_response = 13;
}

message HistoryQueryResult {
  // The history data.
  HistoryVisitData history_data = 1;

  // The query that generated this history data.
  HistoryQuery query = 2;
}

// The requested history data.
message HistoryVisitData {
  repeated HistoryVisitItem visit_item = 1;
}

// The data for a single history item.
message HistoryVisitItem {
  string page_title = 1;

  string page_url = 2;

  Timestamp visit_time = 3;

  repeated string passages = 4;
}

message HistoryQuery {
  // The text used to find matching history items.
  // If empty returns all entries matching the other query parameters.
  string query = 1;

  // The number of history items to fetch.
  int32 num_history_visits = 2;

  // The start and end time to fetch history items.
  DateTimeRange history_search_time_range = 3;
}

message DateTimeRange {
  // Both fields are inclusive.
  Timestamp start_time = 1;

  Timestamp end_time = 2;
}

message ModelPrototypingResponse {
  // The series of prompts sent to the model corresponding to the
  // |prototyping_prompts| in the request.
  repeated string model_prompts = 1;

  // The responses from the model corresponding to |model_prompts|.
  repeated string model_responses = 2;

  // Ordered conversation between the user and the model.
  repeated Conversation conversations = 3;
}

message ModelPrototypingMetadata {
  // An optional tag to use for querying the record from MQLS logs.
  string logging_tag = 1;

  // An optional description to provide more context about the data uploaded by
  // the user.
  string logging_description = 2;

  // The email address of the user who uploaded the data. This will only be
  // populated for the TVC data upload, this field will not be scrubbed and will
  // be used to identify the user if there are issues with this feature.
  string user_email = 3;
}
